optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -0,0 +1,219 @@
+                .penalize();
+    }
+
+    /**
+     * Spread: Processes of the same service must be serviceLocationSpread out across locations.
+     */
+    private void serviceLocationSpread(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_LOCATION_SPREAD.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrProcessAssignment.class)
+                .groupBy(processAssignment -> processAssignment.getService(),
+                         ConstraintCollectors.countDistinct(processAssignment -> processAssignment.getLocation()))
+                .filter((mrService, distinctLocationCount) -> mrService.getLocationSpread() > distinctLocationCount)
+                .penalize();
+    }
+
+    /**
+     * Dependency: The processes of a service depending on another service must run in the neighborhood of a process
+     * of the other service.
+     */
+    private void serviceDependency(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.SERVICE_DEPENDENCY.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrServiceDependency.class)
+                .join(MrProcessAssignment.class,
+                      Joiners.equalTo(MrServiceDependency::getFromService, MrProcessAssignment::getService))
+                .join(MrProcessAssignment.class,
+                      Joiners.equalTo((serviceDependency, processAssignment) -> serviceDependency.getToService(), MrProcessAssignment::getService))
+                .filter((mrServiceDependency, mrProcessAssignmentFrom, mrProcessAssignmentTo) ->
+                                !mrProcessAssignmentFrom.getNeighborhood().equals(mrProcessAssignmentTo.getNeighborhood()))
+                .penalize();
+    }
+
+    /**
+     * Transient usage: Some resources are transient and count towards the maximum capacity of both the original
+     * machine as the newly assigned machine.
+     */
+    private void transientUsage(ConstraintFactory constraintFactory) {
+        Constraint constraint = constraintFactory.newConstraintWithWeight(MrConstraintName.TRANSIENT_USAGE.getName(),
+                                                                          HardSoftLongScore.ofHard(1L));
+        constraint.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      Joiners.equalTo(MrMachineCapacity::getMachine, MrProcessAssignment::getOriginalMachine)
+                )
+                .filter(
+                        (mrMachineCapacity, mrProcessAssignment) -> mrProcessAssignment.isMoved() && mrMachineCapacity.isTransientlyConsumed()
```

Comment:
soft suggestion: For readability, I 'd remove the \n after filter( and add one before &&